### PR TITLE
Don't calculate dataset size by hand, use tf.errors.OutOfRangeError

### PR DESCRIPTION
**Pull request by reuben**

### CODE REVIEW

**Comment by reuben** [comment]
Commit 8053548 is only partially related to the dataset changes, but I'm including here since you're also the relevant reviewer. The motivation for removing `tf.train.Coordinator` is that it swallows every exception inside the `with coord.stop_on_exception()` block, including SyntaxError, DivisionByZero, etc. This causes the silent errors like shown here: https://github.com/mozilla/DeepSpeech/issues/2014

After the tf.data PR we no longer manage threads ourselves, so the coordinator is not used. I just check for KeyboardInterrupt directly so one is able to stop the training but still test with ctrl-c.

File: DeepSpeech.py [hunk 1]
```diff
@@ -447,61 +447,74 @@ def train():
...
+                    return progressbar.widgets.FormatLabel.__call__(self, progress, data)
+
             if FLAGS.show_progressbar:
-                pbar = progressbar.ProgressBar(max_value=num_steps, redirect_stdout=True).start()
-            else:
-                pbar = lambda i: i
+                pbar = progressbar.ProgressBar(widgets=['Epoch {}'.format(epoch),
+                                                        ' | ', progressbar.widgets.Timer(),
+                                                        ' | Steps: ', progressbar.widgets.Counter(),
+                                                        ' | ', LossWidget()])
```
Comment (tilmankamp) [inline DeepSpeech.py]: Nice!

File: DeepSpeech.py [hunk 2]
```diff
@@ -447,61 +447,74 @@ def train():
...
-        with coord.stop_on_exception():
+        try:
             for epoch in range(FLAGS.epochs):
-                if coord.should_stop():
-                    break
-
                 # Training
-                log_info('Training epoch %d...' % epoch)
-                train_loss = run_set('train', train_init_op, train_batches)
-                log_info('Finished training epoch %d - loss: %f' % (epoch, train_loss))
```
Comment (tilmankamp) [inline DeepSpeech.py]: There is one little problem in relying on the progress bar (and I like the beauty of the concept): The huge number of updates leads to huge (slowly downloading) logs and so I was more than one time tempted to disable it.
How about either
 - (dynamically) lowering update frequency or
 - keeping fallback logging if disabled?

File: DeepSpeech.py [hunk 3]
```diff
@@ -447,61 +447,74 @@ def train():
...
 
             # Batch loop
-            for step_index in pbar(range(num_steps)):
-                if coord.should_stop():
+            while True:
+                try:
+                    _, current_step, batch_loss, step_summary = \
+                        session.run([train_op, global_step, loss, step_summaries_op],
+                                    feed_dict=feed_dict)
+                except tf.errors.OutOfRangeError:
```
Comment (tilmankamp) [inline DeepSpeech.py]: "You shall not rely on exception handling for program flow control!" - Honestly: I am totally OK with it. Actually I like it.

File: DeepSpeech.py (hunk 2 above)
Comment (reuben) [reply DeepSpeech.py]: I totally overlooked that there’s a flag to disable the progress bar, I’ll keep the logging if the bar is disabled.

File: DeepSpeech.py (hunk 3 above)
Comment (reuben) [reply DeepSpeech.py]: Yeah, it's unfortunate. When we have eager execution this will be:

```python
for batch in dataset:
    # ...
```

**Pull request merged**